# Prueba Intertrimestral

**Nombre: Victor**

**Apellidos: Ovejero de la Cerda**

**Tiempo de la prueba: 2 Horas**

**Asignatura:** Desarrollo de Aplicaciones para la Visualización de Datos

**Fecha:** 18 de octubre de 2023

**Instrucciones:**

- Escribe código limpio y autoexplicativo.

- Se eliminará 0.5 puntos por usar Seaborn o Matplotlib.

- Se pueden utilizar los materiales de clase.

- Se puede utilizar internet para búsqueda de dudas y documentación.

- No se puede utilizar ningún tipo de LLM.

- No se puede utilizar mensajería instantánea.

- Sube tus resultados a tu repositorio de Github.

- Imprime una versión en PDF en A3 y Portrait del notebook.

- Envialo tus resultados a dmartincorral@icai.comillas.edu adjuntando el PDF y la url del notebook subido al repositorio de Github.


## Inicialización de librerías

Carga aquí todas las librerías que vayas a utilizar.

In [173]:
import pandas as pd
import numpy as np
import sklearn.datasets
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objects as go
import plotly.express as px
import statsmodels.api as sm
import math
import plotly.figure_factory as ff


from sklearn.model_selection import train_test_split
from sklearn.svm import LinearSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import (LogisticRegression,LinearRegression)
from sklearn.ensemble import RandomForestRegressor



from sklearn.metrics import (classification_report, mean_squared_error, mean_absolute_error, mean_absolute_percentage_error,
                            silhouette_score, confusion_matrix, roc_curve, roc_auc_score, precision_recall_curve, r2_score)
from sklearn.linear_model import ElasticNet
from sklearn.cluster import KMeans

# PARA REDUCCION DE DIMENSION (PCA)
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

## Ejercicio 1 (2 puntos):

a) Crea una función que calcule y devuelva el factorial de un número entero. **(0.6 puntos)**

b) Crea una función que verifique si un número es primo o no. **(0.6 puntos)**

c) Muestra en un dataframe los 50 primeros números positivos, si es primo y su factorial utilizando las funciones anteriores. **(0.6 puntos)**

d) ¿Cómo se podría programar en una clase las tres operaciones anteriores? **(0.2 puntos)**

In [24]:
#A

def factorial(x):    
    if(x < 0):
        return "Enter a value bigger or equal to 0."
    else:
        return math.factorial(x)

print(factorial(10))

3628800


In [40]:
#B

def is_prime(n):
    if n <= 1:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False

    # Comprobar si es divisible por cualquier numero (desde 3 hasta la raiz cuadrada de n)
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return False

    return True

print(is_prime(53982894593057))

True


In [64]:
#C

#creamos array
array = list(range(1,51))

#creamos dataframe
df = pd.DataFrame(array,columns=["Numbers"])
df["isPrime"] = False
df["Factorial"] = 0
#Comprobamos cada número para ver si es primo y calcular su factorial.
for i in df["Numbers"]:
    prime = is_prime(i)
    fact = factorial(i)
    # to handle large numbers
    fact = np.array(fact)
    df.iloc[i-1,1] = prime
    df.iloc[i-1,2] = fact    

## Ejercicio 2 (4 puntos):

a) Extrae de sklearn el conjunto de datos __California Housing dataset__
 y transfórmalo a dataframe de pandas **(0.25 puntos)**

b) Construye una función que muestra la estructura del dataset, el número de NAs, tipos de variables y estadísticas básicas de cada una de las variables. **(0.5 puntos)**

c) Construye una __Regresión lineal__ y un __Random forest__ que predigan el __Median house value__ según los datos disponibles. **(0.75 puntos)**

d) Visualiza cuales son las variables (coeficientes) más importantes en cada uno de los modelos. **(1.25 puntos)**

e) Decide a través de las métricas que consideres oportunas, cuál de los dos modelos es mejor, por qué y explica el proceso que has realizado para responder en los puntos anteriores. **(1.25 puntos)**

In [89]:
#A
dataset = sklearn.datasets.fetch_california_housing()
df = pd.DataFrame(data = dataset['data'], columns = dataset['feature_names'])
df["target"] = dataset["target"]

df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422

In [87]:
#B
def estructura(input_df):
    if isinstance(input_df, pd.DataFrame):
        print("Número de NAs: ",input_df.isna().sum().sum())
        print("\n\nTipos de Variables en el dataframe: ", input_df.dtypes)
        print("\n\nEstadística básica del dataframe:\n\n ",input_df.describe())
    else:
        print("El input no es un Dataframe de pandas.")

        
    
    
# estructura(df)

In [93]:
#C

#Train y Test
X = dataset["data"]
y = dataset["target"]
X_train, X_test, y_train, y_test = train_test_split(X, y ,test_size = 0.25, random_state = 123)

In [104]:
#Modelo de Linear Regression
lin_reg = LinearRegression()
lin_reg.fit(X_train,y_train)
print("Linear Regression Score: ",lin_reg.score(X, y))
lr_coef_importance = lin_reg.coef_

Linear Regression Score:  0.6061518392846603


In [105]:
#Modelo de Random Forest
rand_forest= RandomForestRegressor()
rand_forest.fit(X_train,y_train)
print("Random Forest Score: ", rand_forest.score(X, y))
rf_coef_importance = rand_forest.feature_importances_

Random Forest Score:  0.9331311693178401


In [118]:
#D - Visualiza cuales son las variables (coeficientes) más importantes en cada uno de los modelos. (1.25 puntos)

# Visualize Linear Regression Coef.

data = [
    go.Bar(
        x = df.columns,
        y = lr_coef_importance,
        width= np.repeat(0.8,len(df.columns)) # Cambiar la anchura de cada barra

    )
]

layout = go.Layout(barmode = "group",title = "Importance of Variables", xaxis_title="Variables", yaxis_title="Coeficients")

fig = go.Figure(data = data, layout = layout)
fig.show()

In [119]:
# Visualize Random Forest Coef.

data = [
    go.Bar(
        x = df.columns,
        y = rf_coef_importance,
        width= np.repeat(0.8,len(df.columns)) # Cambiar la anchura de cada barra

    )
]

layout = go.Layout(barmode = "group",title = "Importance of Variables", xaxis_title="Variables", yaxis_title="Coeficients")

fig = go.Figure(data = data, layout = layout)
fig.show()

In [124]:
# E - Decide a través de las métricas que consideres oportunas, cuál de los dos modelos es mejor, 
# por qué y explica el proceso que has realizado para responder en los puntos anteriores. (1.25 puntos)

#TRAIN
rf_prediction = rand_forest.predict(X_train)
lr_prediction = lin_reg.predict(X_train)

rf_error = mean_squared_error(y_train,rf_prediction)
lr_error = mean_squared_error(y_train,lr_prediction)
print("RMSE of Random Forest (Train): ", rf_error)
print("RMSE of Linear Regression (Train): ", lr_error)

#TEST
rf_prediction_test = rand_forest.predict(X_test)
lr_prediction_test = lin_reg.predict(X_test)

rf_error_test = mean_squared_error(y_test,rf_prediction_test)
lr_error_test = mean_squared_error(y_test,lr_prediction_test)
print("\nRMSE of Random Forest (Test): ", rf_error_test)
print("RMSE of Linear Regression (Test): ", lr_error_test)

RMSE of Random Forest (Train):  0.03679513960049927
RMSE of Linear Regression (Train):  0.5271385909814755

RMSE of Random Forest (Test):  0.2457714274423007
RMSE of Linear Regression (Test):  0.5162987734042872


Respondiendo al punto E. Se han calculado varias métricas para conocer que modelo es mejor, y se ha probado el modelo sobre el conjunto de testing para poder conocer los errores de precicción.

En primer lugar, se ha calculado el R^2 de ambos modelos, obteniendo 0.93 y 0.61 en el Random Forest y en el Linear Regression respectivamente. Esto, a prioris, muestra que el Random Forest claramente es mejor, explicando el 93% de la variable a diferencia del 61% que explica el Linear Regressor. De todos modos, esto también puede ser dado a un overfitting del modelo, por lo que es necesario probar sobre el conjunto de test.

El RMSE en el conjunto de entrenamiento es mucho mas bajo el del Random Forest. Pero esto se ha mantenido a su vez en el conjunto de test. Obteniendo un RMSE de 0.246 por parte del Random Forest y un 0.516 por parte del Lineal Regressor, mostranso como el random forest ha resultado ser un mejor modelo en este caso.


## Ejercicio 3 (4 puntos):


*Consideremos* el dataset que contiene __The Most Streamed Spotify Songs 2023__ que se encuentra en el respositorio.

Información de las variables:

- track_name: Name of the song
- artist(s)_name: Name of the artist(s) of the song
- vartist_count: Number of artists contributing to the song
- released_year: Year when the song was released
- released_month: Month when the song was released
- release_day: Day of the month when the song was released
- in_spotify_playlists: Number of Spotify playlists the song is included in
- in_spotify_charts: Presence and rank of the song on Spotify charts
- streams: Total number of streams on Spotify
- in_apple_playlists: Number of Apple Music playlists the song is included in
- in_apple_charts: Presence and rank of the song on Apple Music charts
- in_deezer_playlists: Number of Deezer playlists the song is included in
- in_deezer_charts: Presence and rank of the song on Deezer charts
- in_shazam_charts: Presence and rank of the song on Shazam charts
- bpm: Beats per minute, a measure of song tempo
- key: Key of the song
- mode: Mode of the song (major or minor)
- danceability_%: Percentage indicating how suitable the song is for dancing
- valence_%: Positivity of the song's musical content
- energy_%: Perceived energy level of the song
- acousticness_%: Amount of acoustic sound in the song
- instrumentalness_%: Amount of instrumental content in the song
- liveness_%: Presence of live performance elements
- speechiness_%: Amount of spoken words in the song

Para las respuestas b, c, d, e, f y g es imperativo acompañarlas respuestas con una visualización.

a) Lee el fichero en formato dataframe, aplica la función del ejercicio 2.b, elimina NAs y convierte a integer si fuera necesario. **(0.25 puntos)**

b) ¿Cuántos artistas únicos hay? **(0.25 puntos)**

c) ¿Cuál es la distribución de reproducciones? **(0.5 puntos)**

d) ¿Existe una diferencia signitificativa en las reproducciones entre las canciones de un solo artista y las de más de uno? **(0.5 puntos)**

e) ¿Cuáles son las propiedades de una canción que mejor correlan con el número de reproducciones de una canción? **(0.5 puntos)**

f) ¿Cuáles son las variables que mejor predicen las canciones que están por encima el percentil 50? **(1 puntos)**

*Nota: Crea una variable binaria (Hit/No Hit) en base a 3.c, crea una regresión logística y visualiza sus coeficientes.*

g) Agrupa los 4 gráficos realizados en uno solo y haz una recomendación a un sello discográfico para producir un nuevo hit. **(1 puntos)**


In [217]:
#A 
df = pd.read_csv("spotify-2023.csv", encoding="ISO-8859-1")

# estructura(df)
df.head()


track_name    artist(s)_name  artist_count  \
0  Seven (feat. Latto) (Explicit Ver.)  Latto, Jung Kook             2   
1                                 LALA       Myke Towers             1   
2                              vampire    Olivia Rodrigo             1   
3                         Cruel Summer      Taylor Swift             1   
4                       WHERE SHE GOES         Bad Bunny             1   

   released_year  released_month  released_day  in_spotify_playlists  \
0           2023               7            14                   553   
1           2023               3            23                  1474   
2           2023               6            30                  1397   
3           2019               8            23                  7858   
4           2023               5            18                  3133   

   in_spotify_charts    streams  in_apple_playlists  ...  bpm key   mode  \
0                147  141381703                  43  ...  125   B  Major   
1                 48  133716286                  48  ...   92  C#  Major   
2                113  140003974                  94  ...  138   F  Major   
3                100  800840817                 116  ...  170   A  Major   
4                 50  303236322                  84  ...  144   A  Minor   

  danceability_%  valence_% energy_% acousticness_%  instrumentalness_%  \
0             80         89       83             31                   0   
1             71         61       74              7                   0   
2             51         32       53             17                   0   
3             55         58       72             11                   0   
4             65         23       80             14                  63   

   liveness_%  speechiness_%  
0           8              4  
1          10              4  
2          31              6  
3          11             15  
4          11              6  

[5 rows x 24 columns]

In [192]:
# Se quitan los NAs
df.dropna(inplace=True)
print("NAs: :",df.isna().sum().sum())

convert_int = ["streams","in_shazam_charts","in_deezer_playlists"]
# Se convierten las columnas "streams", "in_shazam_charts", "in_deezer_playlists" a int.
var1 = 'BPM110KeyAModeMajorDanceability53Valence75Energy69Acousticness7Instrumentalness0Liveness17Speechiness3'

df[df.eq(var1).any(1)]
df.drop(574,inplace=True)


df['streams'] = df['streams'].astype('int')


# df['in_shazam_charts'] = df['in_shazam_charts'].astype('float')
# df['in_deezer_playlists'] = df['in_deezer_playlists'].astype('float')

NAs: : 0


In [193]:
estructura(df)

Número de NAs:  0


Tipos de Variables en el dataframe:  track_name              object
artist(s)_name          object
artist_count             int64
released_year            int64
released_month           int64
released_day             int64
in_spotify_playlists     int64
in_spotify_charts        int64
streams                  int64
in_apple_playlists       int64
in_apple_charts          int64
in_deezer_playlists     object
in_deezer_charts         int64
in_shazam_charts        object
bpm                      int64
key                     object
mode                    object
danceability_%           int64
valence_%                int64
energy_%                 int64
acousticness_%           int64
instrumentalness_%       int64
liveness_%               int64
speechiness_%            int64
dtype: object


Estadística básica del dataframe:

         artist_count  released_year  released_month  released_day  \
count    816.000000     816.000000      816.000000    816.000000   
mean      

In [171]:
#B - ¿Cuántos artistas únicos hay?

print("Total unique artists is: ",df["artist(s)_name"].nunique())


Total unique artists is:  570


In [181]:
#C - ¿Cuál es la distribución de reproducciones?


fig = px.histogram(df, x="streams",
                   nbins = 30)

fig.show()

In [219]:
#D - ¿Existe una diferencia signitificativa en las reproducciones entre las canciones 
# de un solo artista y las de más de uno?

df_taylor = df.loc[df["artist(s)_name"] == "Taylor Swift"]
df_taylor.groupby("track_name").sum()

fig = px.histogram(df_taylor, x="track_name",y="streams",
                   nbins = 30)

fig.show()

# SI! Existe una diferencia notable entre las reproducciones de diferentes canciones de un solo artista.

In [204]:
#E - ¿Cuáles son las propiedades de una canción que mejor correlan con el 
# número de reproducciones de una canción? 
df_model = df.select_dtypes(include='int')
df_model.head()

y = df_model['streams']
X = df_model.loc[:, df_model.columns != 'streams']
X = X.loc[:, X.columns != 'artist_count']
lin_reg = LinearRegression()
lin_reg.fit(X,y)
print("Linear Regression Score: ",lin_reg.score(X, y))
lr_coef_importance = lin_reg.coef_

print(lr_coef_importance)
data = [
    go.Bar(
        x = X.columns,
        y = lr_coef_importance,
        width= np.repeat(0.8,len(X.columns)) # Cambiar la anchura de cada barra

    )
]

layout = go.Layout(barmode = "group",title = "Importance of Variables", xaxis_title="Variables", yaxis_title="Coeficients")

fig = go.Figure(data = data, layout = layout)
fig.show()

Linear Regression Score:  0.7012605219317694
[ 2770684.44483187  3529454.23691061  1942374.17264615    36732.07895243
  2715817.17332543  2621412.04013865  -170396.07514649 -3967870.47202539
  -157699.07349201  -909304.28767698  -426485.5498869  -1706711.4257045
   631444.73703493  -598828.55131894    88777.1704552  -1049649.42907304]


Las variables más significativas son aquellas con un coeficiente mayor (en valor absoluto).

Por lo tanto, podemos categorizar como las métricas más importantes a:

- released_year
- released_month
- released_day
- in_spotify_playlists
- in_apple_playlists
- in_deezer_charts
- energy_%

In [209]:
# F - ¿Cuáles son las variables que mejor predicen las canciones que están por encima el percentil 50? 
import statistics

median = statistics.median(df_model['streams'])

df_model = df_model[(df_model > median).any(1)]
y = df_model['streams']
X = df_model.loc[:, df_model.columns != 'streams']
X = X.loc[:, X.columns != 'artist_count']


lin_reg = LinearRegression()
lin_reg.fit(X,y)
print("Linear Regression Score: ",lin_reg.score(X, y))
lr_coef_importance = lin_reg.coef_

print(lr_coef_importance)
data = [
    go.Bar(
        x = X.columns,
        y = lr_coef_importance,
        width= np.repeat(0.8,len(X.columns)) # Cambiar la anchura de cada barra

    )
]

layout = go.Layout(barmode = "group",title = "Importance of Variables", xaxis_title="Variables", yaxis_title="Coeficients")

fig = go.Figure(data = data, layout = layout)
fig.show()

Linear Regression Score:  0.6291423631022214
[ 5102904.31280858 -4452224.88539109  2144450.79182218    30524.51305385
  4835089.94309066  2459626.79718948  -758083.76332735 -6311462.2383781
  -713415.42794207 -1588653.85959146  -707625.19167335 -1290485.99992214
  1485100.07055171  -778031.9769384    786746.97897347 -1288398.92198755]


Para las canciones que tengan reproducciones por encima de la mediana, las variables importantes son muy parecidas.

Coge mayor importancia la variable "in_deezer_charts" y sorprendentemente la variable "released_month" ahora tiene un coeficiente negativo, siendo inversamente correlacionada.

También ha subido la importancia de "danceability_%".